lib/mainloop.c

```diff
@@ -297,6 +298,27 @@ main_loop_reload_config_initiate(gpointer user_data)
   main_loop_worker_sync_call(main_loop_reload_config_apply, self);
 }
 
+static void
+block_till_workers_exit()
+{
+  gint64 end_time;
+  static GMutex dummy_workers_running_lock = G_STATIC_MUTEX_INIT;
+
+  end_time = g_get_monotonic_time() + 15*G_TIME_SPAN_SECOND;
+  g_mutex_lock(&dummy_workers_running_lock);
+  while (main_loop_workers_running)
+    if (!g_cond_wait_until(&thread_halt_cond, &dummy_workers_running_lock, end_time))
+      {
+        /* timeout has passed. */
+        msg_error("Main thread timed out while waiting workers threads to exit",
+                  evt_tag_int("workers_running", main_loop_workers_running),
```

Comment:
we should include how long we waited and also what would happen next.

btw: this will probably not be printed, as the internal() source is not in operation at this point.

Reply:
Can I resolve the non printing part somehow? Should I use fprintf here writing to stderr instead of msg_error?

Reply:
I'll change this to fprintf.

How long we waited: will be 15s as we get here only when timeout happened. As for what happens next, I'm not sure. I think we can just continue with exiting. Or should we abort? The only root cause I can think of is somehow the accounting of main_loop_workers_running is messed up (nonzero). But for sure all workers will be exited: we only get here when there are no jobs, and after 10 seconds every thread is forced to exit, so waiting 15seconds and continue is safe imo.
For text:
 fprintf(stderr, "Main thread timed out (15s) while waiting workers threads to exit. workers_running: %d. Continuing ...\n", main_loop_workers_running);